www/js/app.js

```diff
@@ -72,13 +66,25 @@ define(['jquery', 'zimArchiveLoader', 'uiUtil', 'settingsStore','abstractFilesys
     params['showUIAnimations'] = settingsStore.getItem('showUIAnimations') ? settingsStore.getItem('showUIAnimations') === 'true' : true;
     document.getElementById('hideActiveContentWarningCheck').checked = params.hideActiveContentWarning;
     document.getElementById('showUIAnimationsCheck').checked = params.showUIAnimations;
+    // Maximum number of article titles to return (range is 5 - 50, default 25)
+    params['maxSearchResultsSize'] = settingsStore.getItem('maxSearchResultsSize') || 25;
+    document.getElementById('titleSearchRange').value = params.maxSearchResultsSize;
+    document.getElementById('titleSearchRangeVal').innerHTML = params.maxSearchResultsSize;
     // A global parameter that turns caching on or off and deletes the cache (it defaults to true unless explicitly turned off in UI)
     params['useCache'] = settingsStore.getItem('useCache') !== 'false';
     // A parameter to set the app theme and, if necessary, the CSS theme for article content (defaults to 'light')
     params['appTheme'] = settingsStore.getItem('appTheme') || 'light'; // Currently implemented: light|dark|dark_invert|dark_mwInvert
     document.getElementById('appThemeSelect').value = params.appTheme;
     uiUtil.applyAppTheme(params.appTheme);
 
+    // Define global state
+    // An array to hold searches and their state (allows modules to tell which search is current and cancel old searches)
```

Comment:
@mossroy The problem as I see it is that we immediately need to set a flag on the currently running search _from the UI_ (i.e. from `app.js`) to indicate that it has been cancelled, and that flag needs to be immediately available to the subroutines in zimArchive that are operating on that search so they can stop initiating new Promise loops. A new search can be initiated with anew state of 'init', even while the previous search is cancelling. There are queued up Promises which will keep returning for a while, because there is no way to cancel an a queued Promise. So I thought I needed to keep an open handle on the search until we can be sure all Promises have returned, and the returning functions need to see that the state of the old search is 'cancelled', even though the new search has a different state. The problem here is the async nature of search. Only when all searches are definitively over can we prune the list of searches.

I don't have a technical understanding of "state" or how to manage it. Hence my solution here is probably unorthodox...

Reply:
I fully agree with the need to keep a global variable, for the reasons you explained.
What I would suggest to investigate is to keep only the latest search in this global variable, instead of a list of searches. Something like : 

    state['latestSearch'] = { 'prefix': '', 'state': '', 'type': '' };

Instead of passing the current search as a parameter to some functions, these functions could directly read/modify this global variable.
These functions usually need to check that _their_ search has not been cancelled by another one. They could do that by checking `mySearchPrefix !== state.latestSearch.prefix` : if the search prefix has been changed, it means that their search can be aborted.

Reply:
OK, thanks for the explanation. I'll try.

Reply:
@mossroy You were right. There was no need to keep a list of searches. The search cancellation propagates through to the local searches in zimArchive and it seems to work just fine. It's a bit of a mystery to me which objects become local copies, independently modifiable, and which are merely pointers to a global object.